Sources/SwiftInspectorVisitors/Tests/TypeDescriptionSpec.swift

```diff
@@ -0,0 +1,383 @@
+// Created by Dan Federman on 2/2/21.
+//
+// Copyright © 2021 Dan Federman
+//
+// Distributed under the MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import Foundation
+import Nimble
+import Quick
+import SwiftInspectorTestHelpers
+import SwiftSyntax
+
+@testable import SwiftInspectorVisitors
+
+final class TypeDescriptionSpec: QuickSpec {
+
+  override func spec() {
+    describe("init(from decoder: Decoder)") {
+      let encoder = JSONEncoder()
+      let decoder = JSONDecoder()
+      var data: Data!
+      var sut: TypeDescription!
+
+      context("when decoding a simple type") {
+        beforeEach {
+          sut = .simple(name: "Foo")
+          data = try? encoder.encode(sut)
```

Comment:
I am curious what @fdiaz thinks but I feel like these tests would be more powerful if we wrote out the JSON for the decoding tests rather than creating it by the encode method (and vice versa). That way we guard against symmetrical bugs in our encode/decode logic.

Reply:
I'm going to hold off on reviewing this file in detail right now since I feel like it would be good to align on this first.

Reply:
I agree with Michael here although we don't need a JSON, we can pass Data. So we could simply convert from a String to Data using `.data(using: .utf8)` and then pass it to a JSONDecoder / Encoder.

As is, this is too tied to the implementation of this type IMHO so these tests are not as great as they could be.

Reply:
so, I think I've done what y'all suggested in b9bb7ed151f4ff3d3ab3e823255f6f47d163809a. But I truly do not understand how this is more likely to catch a failure or is more maintainable vs what I had.

Reply:
I'm iterating on top of this in another branch to better handle generics. This branch requires changing the shape of this enum somewhat. This change has made it _much_ harder to ship that change. I am very curious what the benefit of this change is 😅

Reply:
Concrete example of the types of changes are [here](https://github.com/fdiaz/SwiftInspector/pull/71/files#diff-e905411ed579d068d03cacb0e10f43774cdfdee6b3bcb046630b740d0866a6ea)

Reply:
From my point of view, the benefits are:
- Writing out the strings prevents against symmetrical bugs. So for example, let's say that you updated `encode(...)` and `decode(...)` to both interact with the wrong key. The old version of this test would pass. The new version wouldn't since the expected data is decoupled from the implementation.
- The new version of the tests are better documenting. I can read the tests and I learn more about how this type works. In the previous world I couldn't do that.

From the perspective of a reviewer of future changes, it can be nice to see how code changes affect real encoded JSON, which we now get. But I see how it's a pain from an authorship POV 😓 

Curious to get @fdiaz 's thoughts as he is deeper in testing best practices than I am.

Reply:
> Writing out the strings prevents against symmetrical bugs. So for example, let's say that you updated encode(...) and decode(...) to both interact with the wrong key. 

Pushing back on this a little bit: what does it mean for a key to be "wrong" in this context? We aren't interacting with a 3rd-party API where we need to abide by a strict contract – if we were, then I'd agree starting with JSON or Data is the right move. I'm pretty sure all we care about here is that we can encode & persist data and then decode that same data back into the original form.

> The new version of the tests are better documenting. I can read the tests and I learn more about how this type works. In the previous world I couldn't do that.

Entirely fair, though I'd wager the investment required to maintain these tests as-is isn't necessarily worth that win.

Also curious to hear Francisco's thoughts 🙂

Reply:
We discussed offline. What we want are two sets of tests:

1. Tests that ensure that we haven't accidentally made a breaking change in our encoded formatting.
2. Tests that ensure that we can decode the same value we encoded.

I've addressed this feedback in a4a42c57df9737712d1209df71a36571b4012388 in #71

Reply:
Thanks for iterating here @dfed. I wish we were able to land on the solution more quickly. I appreciate your patience.

Reply:
And I appreciate 1-day turnaround for reviews on massive PRs. We're moving quickly but landing in a good place. Takes a village.